src/hotspot/share/runtime/trimNativeHeap.cpp

```diff
@@ -0,0 +1,229 @@
+  uint16_t _suspend_count;
+
+  // Statistics
+  unsigned _num_trims_performed;
+
+  bool is_suspended() const {
+    assert(_lock->is_locked(), "Must be");
+    return _suspend_count > 0;
+  }
+
+  uint16_t inc_suspend_count() {
+    assert(_lock->is_locked(), "Must be");
+    assert(_suspend_count < UINT16_MAX, "Sanity");
+    return ++_suspend_count;
+  }
+
+  uint16_t dec_suspend_count() {
+    assert(_lock->is_locked(), "Must be");
+    assert(_suspend_count != 0, "Sanity");
+    return --_suspend_count;
+  }
+
+  bool is_stopped() const {
+    assert(_lock->is_locked(), "Must be");
+    return _stop;
+  }
+
+  bool at_or_nearing_safepoint() const {
+    return SafepointSynchronize::is_at_safepoint() ||
+        SafepointSynchronize::is_synchronizing();
+  }
+  static constexpr int64_t safepoint_poll_ms = 250;
+
+  // in seconds
+  static double now() { return os::elapsedTime(); }
+  static double to_ms(double seconds) { return seconds * 1000.0; }
+
+  struct LogStartStopMark {
+    void log(const char* s) { log_info(trimnative)("NativeTrimmer %s.", s); }
+    LogStartStopMark()  { log("start"); }
+    ~LogStartStopMark() { log("stop"); }
+  };
+
+  void run() override {
+    assert(NativeHeapTrimmer::enabled(), "Only call if enabled");
+
+    LogStartStopMark logStartStop;
+
+    const double interval_secs = (double)TrimNativeHeapInterval / 1000;
+
+    for (;;) {
+      double tnow = now();
+      double next_trim_time = tnow + interval_secs;
+
+      {
+        MonitorLocker ml(_lock, Mutex::_no_safepoint_check_flag);
+        if (_stop) return;
+
+        do { // handle spurious wakeups
+          if (is_suspended()) {
+            ml.wait(0); // infinite
+          } else if (next_trim_time > tnow) {
+            const int64_t wait_ms = MAX2(1.0, to_ms(next_trim_time - tnow));
+            ml.wait(wait_ms);
+          } else if (at_or_nearing_safepoint()) {
+            const int64_t wait_ms = MIN2((int64_t)TrimNativeHeapInterval, safepoint_poll_ms);
+            ml.wait(safepoint_poll_ms);
+          }
+
+          if (_stop) return;
+
+          tnow = now();
+
+        } while (at_or_nearing_safepoint() || is_suspended() || next_trim_time > tnow);
+      } // Lock scope
+
+      // 2 - Trim outside of lock protection.
+      execute_trim_and_log(tnow);
+    } // end for(;;)
+  }
+
+  // Execute the native trim, log results.
+  void execute_trim_and_log(double t1) {
+    assert(os::can_trim_native_heap(), "Unexpected");
+
+    os::size_change_t sc;
+    LogTarget(Info, trimnative) lt;
+    const bool logging_enabled = lt.is_enabled();
+
+    // We only collect size change information if we are logging; save the access to procfs otherwise.
+    if (os::trim_native_heap(logging_enabled ? &sc : nullptr)) {
+      _num_trims_performed++;
+      if (logging_enabled) {
+        double t2 = now();
+        if (sc.after != SIZE_MAX) {
+          const size_t delta = sc.after < sc.before ? (sc.before - sc.after) : (sc.after - sc.before);
```

Comment:
@tstuefe under what situations can `sc.after` be more than `sc.before` after trimming? Is it to handle the case where memory allocations happened in-between the malloc_trim() and the calls to get process memory?

Reply:
Yes. The numbers we print out are RSS; there is no way to get the exact figure for how much memory has been released by the glibc (not even itself knows). RSS is influenced by many factors, and its a rough outline only.